### PR TITLE
Chapter review

### PR DESCRIPTION
**Pull request by dlanger**

### CODE REVIEW

File: contingent/chapter.rst [hunk 1]
```diff
@@ -0,0 +1,1344 @@
...
+     body {font-size: 1.3em; margin: 1em;}
+     div.document {margin: 0em auto; max-width: 40em;}
+     pre {background-color: #eee; padding: 0.6em;}
+   </style>
+
+==========================================
+ Contingent: A Fully Dynamic Build System
+==========================================
+
+Systems to rebuild formatted documents from source texts
```
Comment (dlanger) [inline contingent/chapter.rst]: Throw in a sentence explaining to the reader where these systems are used?

File: contingent/chapter.rst [hunk 2]
```diff
@@ -0,0 +1,1344 @@
...
+      install.rst
+      tutorial.rst
+      api.rst
+
+This list of chapter filenames
+tells Sphinx to include a link to each chapter
+when it builds the ``index.html`` output file.
+It will also include links to any sections within each chapter.
+Stripped of its markup,
+the text that results from the above ``toctree`` command might be::
```
Comment (dlanger) [inline contingent/chapter.rst]: I see that the string "toctree" exists above, but I've not heard of the command `toctree` or what it does.

File: contingent/chapter.rst [hunk 3]
```diff
@@ -0,0 +1,1344 @@
...
+
+This table of contents is a mash-up
+of information from four different files.
+While its basic order and structure come from ``index.rst``,
+the actual title of each chapter and section
+is pulled from the three chapter source files themselves.
+
+If you later reconsider the tutorial’s chapter title —
+after all, the word “newcomer” sounds so antique,
+as if your users are settlers who have just arrived in pioneer Wyoming —
```
Comment (dlanger) [inline contingent/chapter.rst]: Love it.

File: contingent/chapter.rst [hunk 4]
```diff
@@ -0,0 +1,1344 @@
...
+instead of leaving the output in an inconsistent state.
+
+The result would be a system like the old static ``make`` tool,
+but which learned the dependencies between files as they were built —
+that added and removed dependencies dynamically
+as cross references were added, updated, and then later deleted.
+
+In the sections that follow we will construct such a tool in Python,
+named Contingent,
+that guarantees correctness in the presence of dynamic dependencies
```
Comment (dlanger) [inline contingent/chapter.rst]: "Guarantees" is a strong word - @DLTODO come back to this.

File: contingent/chapter.rst [hunk 5]
```diff
@@ -0,0 +1,1344 @@
...
+By eliminating every copy of every intermediate or output asset,
+the build is forced to start over again completely
+with nothing cached — with no memory of its earlier states
+that could possibly lead to a stale product.
+A hefty ``rm`` ``-r`` solves the problem
+of guaranteeing consistency before publishing your Sphinx documentation!
+
+But could we develop a better approach?
+
+What if your build system were a persistent process
```
Comment (dlanger) [inline contingent/chapter.rst]: Are you highlighting "persistent' here because that's how you've chosen to fix this, or because it's not possible in a non-persistent program?

File: contingent/chapter.rst [hunk 6]
```diff
@@ -0,0 +1,1344 @@
...
+The Proper Use of Classes
+=========================
+
+You may have been surprised
+by the absence of classes in the above discussion
+of Python data structures.
+After all, they are a frequent mechanism for organizing data
+and get defined through a dedicated Python syntax.
+
+But it turns out that classes are often
```
Comment (dlanger) [inline contingent/chapter.rst]: I don't understand this sentence.

File: contingent/chapter.rst [hunk 7]
```diff
@@ -0,0 +1,1344 @@
...
+    :start-line: 4
+    :end-line: 6
+
+.. include:: contingent/graphlib.py
+    :code: python
+    :start-line: 16
+    :end-line: 19
+
+The leading underscore
+in front of the attribute names ``_inputs_of`` and ``_consequences_of``
```
Comment (dlanger) [inline contingent/chapter.rst]: Feels like we're missing a code segment here? I've not seen either of these attributes yet.

File: contingent/chapter.rst [hunk 8]
```diff
@@ -0,0 +1,1344 @@
...
+But using a class instance does not really change
+any of the questions we faced above when doing data design:
+it just provides a prettier and less anonymous container.
+
+The true value of classes, then,
+is not that they change the science of data design.
+The value of classes
+is that they let you *hide* your data design from the rest of a program!
+
+From the outside, code can simply ask for a new ``Graph`` instance:
```
Comment (dlanger) [inline contingent/chapter.rst]: This transition feels jarring - maybe a bride sentence about how an example of using classes to hide data design is creating a `Graph` class to encapsulate our dependencies?

File: contingent/chapter.rst [hunk 9]
```diff
@@ -0,0 +1,1344 @@
...
+
+Figure 1 ignored one of the most important relationships
+that we discovered in the opening section of our chapter:
+the way that document titles appear in the table of contents.
+Let’s fill in this detail.
+We will create a node for each title string
+that needs to be generated by parsing an input file
+and then passed to one of our other routines:
+
+>>> g.add_edge('api.rst', 'api-title')
```
Comment (dlanger) [inline contingent/chapter.rst]: I think this could be expanded a bit. The idea of creating an intermediate node to hold how an edge should be displayed - and that that just works, given how the rest of the system is built - is pretty novel, and something the reader would almost certainly not have thought of.

File: contingent/chapter.rst [hunk 10]
```diff
@@ -0,0 +1,1344 @@
...
+
+Learning Connections
+====================
+
+We now have a way for Contingent
+to keep track of tasks and the relationships between them.
+If we look more closely at Figure 2, however,
+we see that it is actually a little hand wavy and vague:
+what does it mean to say that
+``index.rst``, ``api-title``, and ``tutorial.html`` are “tasks?”
```
Comment (dlanger) [inline contingent/chapter.rst]: Where did we say this?

File: contingent/chapter.rst [hunk 11]
```diff
@@ -0,0 +1,1344 @@
...
+enclosed in back ticks,
+which on output are replaced with the title
+from the corresponding document in the output.
+
+Here is the content of our example
+``index.txt``, ``api.txt``, and ``tutorial.txt``,
+illustrating titles, document bodies, and cross-references
+from our little document format:
+
+>>> index = """
```
Comment (dlanger) [inline contingent/chapter.rst]: Showing these as if you're in a REPL feels like it's just complicating things - when we showed files earlier, they weren't in a REPL.

File: contingent/chapter.rst [hunk 12]
```diff
@@ -0,0 +1,1344 @@
...
+and applying one or more view transformations
+to convert the internal representation to its output form.
+
+Contingent manages tasks by grouping them into a ``Project``,
+a sort of build system busybody
+that injects itself into the middle of the build process,
+noting every time one task talks to another
+to construct a graph of the relationships between all the tasks.
+
+>>> from contingent.projectlib import Project, Task
```
Comment (dlanger) [inline contingent/chapter.rst]: Why the `Task` import?

File: contingent/chapter.rst (hunk 11 above)
Comment (dlanger) [inline contingent/chapter.rst]: This makes more sense later on.

File: contingent/chapter.rst [hunk 13]
```diff
@@ -0,0 +1,1344 @@
...
+  If you are coming from an orthodox object-oriented tradition,
+  this function-oriented design may look a little weird.
+  In an OO solution,
+  ``parse()`` would return some sort of ``Document`` object
+  that has ``title_of()`` as a method or property.
+  In fact, Sphinx works exactly this way.
+  Its ``Parser`` subsystem producing a “Docutils document tree” object
+  for the other parts of the system.
+
+  Contingent is not opinionated
```
Comment (dlanger) [inline contingent/chapter.rst]: Will there be any examples/documentation on how to use the other? Feels like I'm being left dangling.

File: contingent/chapter.rst [hunk 14]
```diff
@@ -0,0 +1,1344 @@
...
+that transitively connects all the tasks
+required to produce the output,
+from reading the input file,
+parsing and transforming the document,
+and rendering the result:
+
+..
+ >>> open('figure3.dot', 'w').write(as_graphviz(project._graph)) and None
+
+.. image:: figure3.png
```
Comment (dlanger) [inline contingent/chapter.rst]: No label on this figure.

File: contingent/chapter.rst [hunk 15]
```diff
@@ -0,0 +1,1344 @@
...
+   will not be able to tell that they have not simply invoked
+   the plain task function itself.
+
+Steps 4 and 5 maintain the task stack itself,
+which is then used by step 2 to perform the consequences tracking
+that is our whole reason for building a task stack in the first place.
+
+Since each task gets surrounded by its own copy of the wrapper function,
+the mere invocation and execution of the normal stack of tasks
+will produce a graph of relationships as an invisible side effect.
```
Comment (dlanger) [inline contingent/chapter.rst]: This is very cool, and could be expanded.

File: contingent/chapter.rst [hunk 16]
```diff
@@ -0,0 +1,1344 @@
...
+The name of the innermost executing method at the moment of the error
+will necessarily direct us to exactly the class, and set,
+involved in the mistake.
+There is no need to create a subclass of ``set``
+for every possible application of the data type,
+so long as we put that conventional underscore in front of data
+structure attributes and then are careful not to touch them
+from code outside of the class.
+
+Contingent demonstrates how crucial the Facade pattern,
```
Comment (dlanger) [inline contingent/chapter.rst]: Maybe shout-out to this earlier when an example comes up?

File: contingent/contingent/io.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+    if _libc is not None:
+        return
+    ctypes.cdll.LoadLibrary('libc.so.6')
+    _libc = ctypes.CDLL('libc.so.6', use_errno=True)
+    _libc.inotify_add_watch.argtypes = [
+        ctypes.c_int, ctypes.c_char_p, ctypes.c_uint32]
+    _libc.inotify_add_watch.restype = ctypes.c_int
+
+def wait_on(paths):
+    # TODO: auto-detect when the OS does not offer libc or libc does not
```
Comment (dlanger) [inline contingent/contingent/io.py]: Remove TODO, or edit comment?

File: contingent/posts/subtraction.ipynb [hunk 18]
```diff
@@ -0,0 +1,45 @@
+{
```
Comment (dlanger) [inline contingent/posts/subtraction.ipynb]: Where does this file fit in with the rest?

File: contingent/chapter.rst [hunk 19]
```diff
@@ -0,0 +1,1344 @@
...
+turning source into markup involves several steps:
+reading the raw text from disk,
+parsing the text to a convenient internal representation,
+processing any directives the author may have specified,
+resolving cross-references or other external dependencies
+(such as include files),
+and applying one or more view transformations
+to convert the internal representation to its output form.
+
+Contingent manages tasks by grouping them into a ``Project``,
```
Comment (dlanger) [inline contingent/chapter.rst]: The stuff around `Project` is some of the most interesting in this chapter, but I found it a little hard to follow - it's introduced almost in passing here, and then we jump back to it periodically. It's got some interesting technical dimensions which I think could be free-standing explored a little, before jumping into how it fits into the whole ecosystem.

File: contingent/chapter.rst [hunk 20]
```diff
@@ -0,0 +1,1344 @@
...
+
+>>> project.start_tracing()
+>>> project.rebuild()
+>>> print(project.stop_tracing())
+calling read('tutorial.txt')
+calling parse('tutorial.txt')
+calling render('tutorial.txt')
+calling title_of('tutorial.txt')
+
+Success!
```
Comment (dlanger) [inline contingent/chapter.rst]: Could you add an example of how to get Contingent going on a real project? You've shown a bunch of REPL examples, but I don't follow how I would run the implementation on my own (with a file watcher, etc.).

**Comment by brandon-rhodes** [comment]
Thank you for all of these comments — we will hopefully be getting to your comments this weekend — look for an update early next week!

File: contingent/chapter.rst (hunk 7 above)
Comment (drocco007) [inline contingent/chapter.rst]: GitHub's reStructuredText renderer doesn't handle `include` directives, and, what's worse, silently collapses them out of its preview. Here is the rendered version; please let us know if this bit is still confusing:

----

Behind the scenes, a pair of dictionaries has already been built to store edges using the logic we outlined in the previous section:

    class Graph:
        """A directed graph of the relationships among build tasks.
    
        def __init__(self):
            self._inputs_of = defaultdict(set)
            self._consequences_of = defaultdict(set)

The leading underscore in front of the attribute names…
